homeassistant/components/vorwerk/vacuum.py

```diff
@@ -0,0 +1,348 @@
+    VORWERK_ROBOTS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(minutes=SCAN_INTERVAL_MINUTES)
+
+SUPPORT_VORWERK = (
+    SUPPORT_BATTERY
+    | SUPPORT_PAUSE
+    | SUPPORT_RETURN_HOME
+    | SUPPORT_STOP
+    | SUPPORT_START
+    | SUPPORT_CLEAN_SPOT
+    | SUPPORT_STATE
+    | SUPPORT_LOCATE
+)
+
+ATTR_CLEAN_START = "clean_start"
+ATTR_CLEAN_STOP = "clean_stop"
+ATTR_CLEAN_AREA = "clean_area"
+ATTR_CLEAN_BATTERY_START = "battery_level_at_clean_start"
+ATTR_CLEAN_BATTERY_END = "battery_level_at_clean_end"
+ATTR_CLEAN_SUSP_COUNT = "clean_suspension_count"
+ATTR_CLEAN_SUSP_TIME = "clean_suspension_time"
+ATTR_CLEAN_PAUSE_TIME = "clean_pause_time"
+ATTR_CLEAN_ERROR_TIME = "clean_error_time"
+ATTR_LAUNCHED_FROM = "launched_from"
+
+ATTR_NAVIGATION = "navigation"
+ATTR_CATEGORY = "category"
+ATTR_ZONE = "zone"
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Vorwerk vacuum with config entry."""
+
+    _LOGGER.debug("Adding vorwerk vacuums")
+    async_add_entities(
+        [
+            VorwerkConnectedVacuum(robot)
+            for robot in hass.data[VORWERK_DOMAIN][entry.entry_id][VORWERK_ROBOTS]
+        ],
+        True,
+    )
+
+    platform = entity_platform.current_platform.get()
+    assert platform is not None
+
+    platform.async_register_entity_service(
+        "custom_cleaning",
+        {
+            vol.Optional(ATTR_MODE, default=2): cv.positive_int,
+            vol.Optional(ATTR_NAVIGATION, default=1): cv.positive_int,
+            vol.Optional(ATTR_CATEGORY, default=4): cv.positive_int,
+            vol.Optional(ATTR_ZONE): cv.string,
+        },
+        "vorwerk_custom_cleaning",
+    )
+
+
+class VorwerkConnectedVacuum(StateVacuumEntity):
+    """Representation of a Vorwerk Connected Vacuum."""
+
+    def __init__(self, robot):
+        """Initialize the Vorwerk Connected Vacuum."""
+        self.robot = robot
+        self._available = False
+        self._name = f"{self.robot.name}"
+        self._robot_has_map = False
+        self._robot_serial = self.robot.serial
+        self._status_state = None
+        self._clean_state = None
+        self._state = None
+        self._clean_time_start = None
+        self._clean_time_stop = None
+        self._clean_area = None
+        self._clean_battery_start = None
+        self._clean_battery_end = None
+        self._clean_susp_charge_count = None
+        self._clean_susp_time = None
+        self._clean_pause_time = None
+        self._clean_error_time = None
+        self._launched_from = None
+        self._battery_level = None
+        self._robot_boundaries = []
+        self._robot_stats = None
+
+    def update(self):
+        """Update the states of Vorwerk Vacuums."""
+        _LOGGER.debug("Running Vorwerk Vacuums update for '%s'", self.entity_id)
+        try:
+            if self._robot_stats is None:
+                self._robot_stats = self.robot.get_general_info().json().get("data")
```

Comment:
Once the other platforms are addd, this probably should move to a seperate class the implements dataupdatecoordinator.

Reply:
I'm new to home assistant integrations could you show me an example?

Reply:
> 
> 
> Once the other platforms are addd, this probably should move to a seperate class the implements dataupdatecoordinator.

I'm new to home assistant integrations could you show me an example?

Reply:
I'm new to home assistant integrations could you show me an example?

Reply:
Docs are here: https://developers.home-assistant.io/docs/integration_fetching_data/ and for instance Synology_DSM uses it: https://github.com/home-assistant/core/blob/dev/homeassistant/components/synology_dsm/__init__.py looks like a fairly complex implementations, but just search for other components that import DataUpdateCoordinator

Reply:
If it's okay, I will change this with the next PR.